src/Token/Builder.php

```diff
@@ -17,16 +17,23 @@
 use function in_array;
 
 /** @immutable */
-final class Builder implements BuilderInterface
+final readonly class Builder implements BuilderInterface
 {
-    /** @var array<non-empty-string, mixed> */
-    private array $headers = ['typ' => 'JWT', 'alg' => null];
-
-    /** @var array<non-empty-string, mixed> */
-    private array $claims = [];
+    /**
+     * @param array<non-empty-string, mixed> $headers
+     * @param array<non-empty-string, mixed> $claims
+     */
+    private function __construct(
+        private Encoder $encoder,
+        private ClaimsFormatter $claimFormatter,
+        private array $headers = ['typ' => 'JWT', 'alg' => null],
+        private array $claims = [],
+    ) {
+    }
 
-    public function __construct(private readonly Encoder $encoder, private readonly ClaimsFormatter $claimFormatter)
+    public static function new(Encoder $encoder, ClaimsFormatter $claimFormatter): self
     {
+        return new self($encoder, $claimFormatter);
     }
```

Comment:
This is something that we'll need to coordinate in v5.5 the migration path for...

Reply:
Migration API added in https://github.com/lcobucci/jwt/pull/1092